### PR TITLE
Make snappy package

### PR DESCRIPTION
**Pull request by nsg**
This PR adds the need bits to build a snap of the client.

http://snapcraft.io/

Will work on any modern (Linux 4.4+) and supported (Ubuntu, Arch, Debian, Gentoo, Fedora, OpenSUSE, ...) distribution.

### CODE REVIEW

**Comment by nsg** [comment]
@petterarvidsson I have thought about this for months but I finally updated my main workstation, so I now have a supported kernel and I spent the day learning how to make this work. What do you think?

**Comment by nsg** [comment]
Note: atm you need to install it with `--devmode` (security enforcement off). When the application is denied access to search for the texture/shader/.. files outside it's jail it gets a little confused:

```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Shader code must not be empty
```

```
Sep 10 23:36:44 narsil kernel: [ 2943.359635] audit: type=1400 audit(1473543404.496:241): apparmor="DENIED" operation="open" profile="snap.konstructs-client.konstructs-client" name="/home/nsg/code/konstructs/client/shaders/chunk.vert" pid=20014 comm="konstructs" requested_mask="r" denied_mask="r" fsuid=1000 ouid=1000
```

**Comment by petterarvidsson** [comment]
@nsg This is super cool! I guess a snap could then also be used on all other distros that support snaps?

How can I install the snap I built locally? I tried:

```
make snap
sudo snap install konstructs-client_6_amd64.snap
```
But the response I get indicates a failure:
```
error: cannot find signatures with metadata for snap "konstructs-client_6_amd64.snap"
```

**Comment by nsg** [comment]
@petterarvidsson yes, exactly

```
nsg@narsil package $ snap install --help
Usage:
  snap [OPTIONS] install [install-OPTIONS] <snap>

The install command installs the named snap in the system.

Application Options:
      --version              print the version and exit

Help Options:
  -h, --help                 Show this help message

[install command options]
          --channel=         Use this channel instead of stable
          --edge             Install from the edge channel
          --beta             Install from the beta channel
          --candidate        Install from the candidate channel
          --stable           Install from the stable channel
          --devmode          Request non-enforcing security
          --jailmode         Override a snap's request for non-enforcing security
          --revision=        Install the given revision of a snap, to which you must have developer access
          --force-dangerous  Install the given snap file even if there are no pre-acknowledged signatures for it, meaning it was not verified and could be
                             dangerous
```

Check out the last option.

**Comment by nsg** [comment]
I have also published it at the store under the edge channel (intended for dev builds and unstable/testing things...). But looks like it needs to be reviewed, probably because it's a new package... a good thing that there are some type of verification.

**Comment by petterarvidsson** [comment]
Interesting, snap installs, but now it fails to load the GL driver. I have tried to google a bit but can not find any given solution.
```
konstructs-client 
libGL error: unable to load driver: i965_dri.so
libGL error: driver pointer missing
libGL error: failed to load driver: i965
libGL error: unable to load driver: swrast_dri.so
libGL error: failed to load driver: swrast
GLFW error 65543: GLX: Failed to create context: GLXBadFBConfig
Caught a fatal error: Could not create an OpenGL 3.3 context!
```

Could there be some access right required?

**Comment by petterarvidsson** [comment]
Also tried with `--devmode`, but with the same problem.

**Comment by nsg** [comment]
Maybe we need to bundle some more packages, I have only added `libgl1-mesa-dri` under `stage-packages:` in `snapcraft.yaml`.

File: package/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      - network
+      - opengl
+      - unity7
+
+parts:
+  client:
+    plugin: cmake
+    source: ..
+    configflags: [-DCMAKE_INSTALL_PREFIX=/]
+    build-packages:
```
Comment (petterarvidsson) [inline package/snapcraft.yaml]: @nsg Tested to build without glew, glfw and glu here, works as good / bad as before so I'd say they can safely be removed. The ones that should be used are the ones that are included in our repository.

**Comment by petterarvidsson** [comment]
Hmm, if I understand correctly, in theory the driver should already be included in the mesa package:
```
dpkg -S /usr/lib/x86_64-linux-gnu/dri/i965_dri.so
libgl1-mesa-dri:amd64: /usr/lib/x86_64-linux-gnu/dri/i965_dri.so
```

I wonder if it just me having this problem if the problem exists for everyone with intel graphics.

**Comment by nsg** [comment]
We can't assume that you have that package installed, the relevant one is:

`/snap/konstructs-client/current/usr/lib/x86_64-linux-gnu/dri/i965_dri.so`

... and the generated wrapper looks like this:
```
$ cat /snap/konstructs-client/current/command-konstructs-client.wrapper 
#!/bin/sh
export PATH="$SNAP/usr/sbin:$SNAP/usr/bin:$SNAP/sbin:$SNAP/bin:$PATH"
export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$SNAP/lib:$SNAP/usr/lib:$SNAP/lib/x86_64-linux-gnu:$SNAP/usr/lib/x86_64-linux-gnu"
export LD_LIBRARY_PATH="$SNAP/usr/lib/nvidia-361:$SNAP/usr/lib/x86_64-linux-gnu:$LD_LIBRARY_PATH"

LD_LIBRARY_PATH=$SNAP_LIBRARY_PATH:$LD_LIBRARY_PATH
exec "$SNAP/konstructs" "$@"
```

**Comment by petterarvidsson** [comment]
@nsg Maybe this is something: https://bugs.launchpad.net/snappy/+bug/1584178 ?

File: package/snapcraft.yaml (hunk 1 above)
Comment (nsg) [inline package/snapcraft.yaml]: can you clarify "The ones that should be used are the ones that are included in our repository"?

**Comment by petterarvidsson** [comment]
@nsg Yes, I got that, but I thought the same deb package was used to populate the snap as was used on my installation?

The file is there also for my snap.

**Comment by nsg** [comment]
Because you built it, yup I guess that's true.

**Comment by nsg** [comment]
@petterarvidsson was you able to reproduce the glxgears problem from the bug linked above?

**Comment by petterarvidsson** [comment]
@nsg I am trying to figure out how to add an ENV to the snap

**Comment by petterarvidsson** [comment]
@nsg had to add a wrapper script but that did help. Do you want me to commit the changes?

**Comment by nsg** [comment]
Well, if it made no difference I see no point to include it :) I will see if I have time to reinstall my intel based laptop (for a more modern kernel) later this week and see if I have the same problem.

**Comment by petterarvidsson** [comment]
@nsg I meant to write that it did help. With the script it worked. It's just that it's not such an elegant solution.

**Comment by petterarvidsson** [comment]
@nsg Pushed the changes. One commit is for removing the unnecessary deps. The second is for adding the env required to make it work on my setup.

**Comment by petterarvidsson** [comment]
Feel free to drop any of them!

**Comment by nsg** [comment]
@petterarvidsson 

I have now updated this PR with a few changes that I liked to do, including adding a desktop entry (it is required for graphical apps).

File: package/snapcraft.yaml (hunk 1 above)
Comment (petterarvidsson) [inline package/snapcraft.yaml]: The ones used included in our repository are the ones that are found in the dependencies folder

**Comment by petterarvidsson** [comment]
@nsg Looking great :+1:

**Pull request merged**